access/handler.go

```diff
@@ -580,6 +581,194 @@ func (h *Handler) GetAccountAtBlockHeight(
+	if err != nil {
+		return nil, err
+	}
+
+	accountKey, err := convert.AccountKeyToMessage(*keyByIndex)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "invalid key: %v", err)
+	}
+
+	return &access.AccountKeyResponse{
+		AccountKey: accountKey,
+		Metadata:   metadata,
+	}, nil
+}
+
+// GetAccountKeysAtLatestBlock returns an account public keys by address at the latest sealed block.
+func (h *Handler) GetAccountKeysAtLatestBlock(
+	ctx context.Context,
+	req *access.GetAccountKeysAtLatestBlockRequest,
+) (*access.AccountKeysResponse, error) {
+	metadata, err := h.buildMetadataResponse()
+	if err != nil {
+		return nil, err
+	}
+
+	address, err := convert.Address(req.GetAddress(), h.chain)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "invalid address: %v", err)
+	}
+
+	accountKeys, err := h.api.GetAccountKeysAtLatestBlock(ctx, address)
+	if err != nil {
+		return nil, err
+	}
+
+	var publicKeys []*entities.AccountKey
+
+	for i, key := range accountKeys {
+		accountKey, err := convert.AccountKeyToMessage(key)
+		if err != nil {
+			return nil, status.Errorf(codes.InvalidArgument, "invalid key: %v", err)
+		}
+
+		publicKeys[i] = accountKey
+	}
+
+	return &access.AccountKeysResponse{
+		AccountKeys: publicKeys,
+		Metadata:    metadata,
+	}, nil
+}
+
+// GetAccountKeyAtBlockHeight returns an account public keys by address and key index at the given block height.
+func (h *Handler) GetAccountKeyAtBlockHeight(
+	ctx context.Context,
+	req *access.GetAccountKeyAtBlockHeightRequest,
+) (*access.AccountKeyResponse, error) {
+	metadata, err := h.buildMetadataResponse()
+	if err != nil {
+		return nil, err
+	}
+
+	address, err := convert.Address(req.GetAddress(), h.chain)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "invalid address: %v", err)
+	}
+
+	keyByIndex, err := h.api.GetAccountKeyAtBlockHeight(ctx, address, uint64(req.GetIndex()), req.GetBlockHeight())
+	if err != nil {
+		return nil, err
+	}
+
+	accountKey, err := convert.AccountKeyToMessage(*keyByIndex)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "invalid key: %v", err)
+	}
+
+	return &access.AccountKeyResponse{
+		AccountKey: accountKey,
+		Metadata:   metadata,
+	}, nil
+}
+
+// GetAccountKeysAtBlockHeight returns an account public keys by address at the given block height.
+func (h *Handler) GetAccountKeysAtBlockHeight(
+	ctx context.Context,
+	req *access.GetAccountKeysAtBlockHeightRequest,
+) (*access.AccountKeysResponse, error) {
+	metadata, err := h.buildMetadataResponse()
+	if err != nil {
+		return nil, err
+	}
+
+	address, err := convert.Address(req.GetAddress(), h.chain)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "invalid address: %v", err)
+	}
+
+	accountKeys, err := h.api.GetAccountKeysAtBlockHeight(ctx, address, req.GetBlockHeight())
+	if err != nil {
+		return nil, err
+	}
+
+	var publicKeys []*entities.AccountKey
```

Comment:
you should either allocate this array, or use append below. this code will panic